python/pyiceberg/catalog/rest.py

```diff
@@ -0,0 +1,373 @@
+        """Constructs the endpoint
+
+        Args:
+            prefixed: If the prefix return by the config needs to be appended
+
+        Returns:
+            The base url of the rest catalog
+        """
+
+        url = self.host
+        url = url if url.endswith("/") else url + "/"
+
+        if prefixed:
+            url += self.config.overrides.get(PREFIX, "")
+            url = url if url.endswith("/") else url + "/"
+
+        return url + endpoint.format(**kwargs)
+
+    def _fetch_access_token(self, token: str) -> TokenResponse:
+        client, secret = self._split_token(token)
+        data = {GRANT_TYPE: CLIENT_CREDENTIALS, CLIENT_ID: client, CLIENT_SECRET: secret, SCOPE: CATALOG_SCOPE}
+        response = requests.post(self.url(Endpoints.get_token, prefixed=False), data=data)
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {401: BadCredentialsError})
+
+        return TokenResponse(**response.json())
+
+    def _fetch_config(self) -> ConfigResponse:
+        response = requests.get(self.url(Endpoints.get_config, prefixed=False), headers=self.headers)
+        response.raise_for_status()
+        return ConfigResponse(**response.json())
+
+    def _split_namespace_and_table(self, identifier: Union[str, Identifier]) -> Properties:
+        identifier = self.identifier_to_tuple(identifier)
+        return {"namespace": NAMESPACE_SEPARATOR.join(identifier[:-1]), "table": identifier[-1]}
+
+    def _split_namespace_and_name(self, identifier: Union[str, Identifier]) -> Dict[str, Union[Tuple[str, ...], str]]:
+        identifier = self.identifier_to_tuple(identifier)
+        return {"namespace": identifier[:-1], "name": identifier[-1]}
+
+    def _handle_non_200_response(self, exc: HTTPError, error_handler: Dict[int, Type[Exception]]):
+        try:
+            response = ErrorResponse(**exc.response.json())
+        except Exception:  # pylint: disable=W0012,W0703
+            # In the case we don't have a proper response
+            response = ErrorResponse(
+                error=ErrorResponseMessage(
+                    message=f"Could not decode json payload: {exc.response.text}",
+                    type="RESTError",
+                    code=exc.response.status_code,
+                )
+            )
+
+        code = exc.response.status_code
+        if code in error_handler:
+            raise error_handler[code](response) from exc
+        elif code == 400:
+            raise BadRequestError(response.error.message) from exc
+        elif code == 401:
+            raise UnauthorizedError(response.error.message) from exc
+        elif code == 503:
+            raise ServiceUnavailableError(response.error.message) from exc
+        elif 500 <= code < 600:
+            raise ServerError(response.error.message) from exc
+        else:
+            raise RESTError(response.error.message) from exc
```

Comment:
I think that this should be `response` so that the entire response context is included in the trace.

Reply:
In the lines above we try to parse the ErrorResponse, if that doesn't work, we just return the plain response:
```
try:
    response = ErrorResponse(**exc.response.json())
except Exception:  # pylint: disable=W0012,W0703
    # In the case we don't have a proper response
    response = ErrorResponse(
        error=ErrorResponseMessage(
            message=f"Could not decode json payload: {exc.response.text}",
            type="RESTError",
            code=exc.response.status_code,
        )
    )
```

Reply:
What I meant was that in the event that we aren't translating to a specific error type, I think it would make sense to pass `type` through from the `ErrorResponse`. This is minor, though.

Reply:
I'm happy to create a mapping from the RESTException as defined in the catalog to our PythonError. We could also do this later to not add to this PR.

Reply:
I think that the current mapping is fine. But if we have a service endpoint that sends something unexpected, it would be better to have the entire error response in the error that gets raised here. As it is right now, this will discard extra information and make it harder to debug if the service is doing something strange.

I see 3 options:
* Add extra data to `RESTError` that gets printed along with it: `raise RESTError(response.error.message, response.error)`
* Serialize the error as a string and use that for the message: `raise RESTError(repr(response.error))`
* Save the original exception and use it as context to make the `type` accessible: `raise RESTError from exc`

We may want to use `from exc` anyway? I'm not sure if that's going to be automatically included as `__cause__`, since this is in a handler function.